source/treap.h

```diff
@@ -16,6 +16,7 @@ struct Treap {
 	Treap(Val x, Treap* l = NULL, Treap *r = NULL) {
 		val = x; sz = 1; prio = rng();
 		left = l; right = r; rev = false; sum = 0;
+		recalc();
 	}
```

Comment:
Looking at the rest of the code, this constructor is only ever called as `Treap(v)`, so we could drop the arguments and have it initialize `left = right = NULL, sum=x`.

```suggestion
	Treap(Val x) {
		val = x; sz = 1; prio = rng();
		left = right = NULL;
		rev = false; sum = val;
	}
```

Reply:
Sounds good. I don't think we'll ever want to create a Node manually with left/right child.

Reply:
And we really shouldn't, building the tree manually we'd need to make sure the priorities are respected